java/oauth-custom-challengehandler/src/main/java/com/esri/arcgisruntime/sample/oauthcustomchallengehandler/MainActivity.java

```diff
@@ -0,0 +1,298 @@
+        sOAuthChallengeCountDownLatch = new OAuthChallengeCountDownLatch(1);
+
+        final String portalUrl = challenge.getRemoteResource().getUri();
+
+        // get the OAuth config for the portalUrl associated with the challenge
+        try {
+          mOAuthConfiguration = AuthenticationManager.getOAuthConfiguration(portalUrl);
+        } catch (MalformedURLException e) {
+          throw new RuntimeException(e.getMessage(), e.getCause());
+        }
+
+        // get the authorization url which loads the OAuth login page to display in the dialog
+        final String url = OAuthTokenCredentialRequest
+            .getAuthorizationUrl(portalUrl, mOAuthConfiguration.getClientId(), mOAuthConfiguration.getRedirectUri(), 0);
+
+        runOnUiThread(new Runnable() {
+          @Override public void run() {
+
+            // setup webview
+            WebView webView = new WebView(MainActivity.this);
+            webView.setLayoutParams(new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
+                ViewGroup.LayoutParams.MATCH_PARENT));
+            WebSettings webSettings = webView.getSettings();
+            webSettings.setBuiltInZoomControls(true);
+            webSettings.setJavaScriptEnabled(true);
+
+            // setup webviewClient
+            CustomWebViewClient customWebViewClient = new CustomWebViewClient(portalUrl,
+                mOAuthConfiguration.getClientId(), mOAuthConfiguration.getRedirectUri());
+            webView.setWebViewClient(customWebViewClient);
+
+            webView.loadUrl(url);
+            // set the webview on the dialog
+            mOAuthLoginDialog.setContentView(webView);
+            // display the dialog
+            mOAuthLoginDialog.show();
+          }
+        });
+
+        try {
+          // Wait for the OAuth browser page to be shown and user to login and for the CustomWebViewClient
+          // to get the auth code and request an OAuthTokenCredential with it
+          sOAuthChallengeCountDownLatch.await();
+        } catch (InterruptedException ie) {
+        }
+        // Get the credential from the latch once it has been counted down
+        OAuthTokenCredential credential = sOAuthChallengeCountDownLatch.getOAuthTokenCredential();
+        if (credential != null) {
+          ret = new AuthenticationChallengeResponse(AuthenticationChallengeResponse.Action.CONTINUE_WITH_CREDENTIAL,
+              credential);
+        }
+        // Reset the latch for the next challenge
+        sOAuthChallengeCountDownLatch = null;
+        // dismiss the dialog
+        mOAuthLoginDialog.dismiss();
+      }
+      return ret;
+    }
+  };
+
+  /**
+   * CustomWebViewClient overrides the shouldOverrideUrlLoading method to handle errors on the login page and to request
+   * OAuthTokenCredential after having received the auth code when the user provides the login credential/
+   */
+  private class CustomWebViewClient extends WebViewClient {
```

Comment:
I suggest moving this class to the bottom, below `finishOAuthChallenge`, `onResume` etc...

Reply:
updated